explainer.md

```diff
@@ -48,218 +49,298 @@ Provide web apps with efficient access to built-in (software and hardware) media
+-   A **VideoEncoder** encodes VideoFrames to produce EncodedVideoChunks.
 
-An **AudioDecoder** is a TransformStream from EncodedAudioChunk to AudioPacket.
+-   An **AudioDecoder** decodes EncodedAudioChunks to produce AudioPackets.
 
-An **AudioTrackWriter** converts a WritableStream of AudioPacket into a MediaStreamTrack.
+-   A **VideoDecoder** decodes EncodedVideoChunks to produce VideoFrames.
 
-A **VideoTrackReader** converts a MediaStreamTrack into a ReadableStream of VideoFrame.
 
-A **VideoEncoder** is a TransformStream from VideoFrame to EncodedVideoChunk.
+WebCodecs will also define mechanisms for importing content from getUserMedia().
 
-A **VideoDecoder** is a TransformStream from EncodedVideoChunk to VideoFrame.
+-   An **AudioTrackReader** converts an audio MediaStreamTrack into a ReadableStream of AudioPacket.
 
-A **VideoTrackWriter** converts a WritableStream of VideoFrame into a MediaStreamTrack.
+-   A **VideoTrackReader** converts a video MediaStreamTrack into a ReadableStream of VideoFrame.
 
 ## Examples
-### Example of video rendering to Canvas for low-latency live streaming or cloud gaming
+### Example of video rendering to Canvas for extremely low-latency streaming (e.g. cloud gaming)
 
 ```javascript
-// App defines a simple WritableSink that renders new frames as soon as they arrive.
-class CanvasRendererSink {
-  constructor(canvas) {
-    this._context = canvas.getContext('bitmaprenderer');
-  }
+// The app provides error handling.
+function onDecoderError(error) { ... }
 
-  write(videoFrame) {
-    _context.transferFromImageBitmap(videoFrame.image);
-    return Promise.resolve();
-  }
-}
+// App provides stream of encoded chunks to decoder.
+function streamEncodedChunks(decodeCallback) { ... }
 
-// The app provides ReadableStreams of encoded video
-// in the form of byte arrays defined by a codec such as vp8
-// (not in a media container such as mp4 or webm).
-const {encodedVideo} = ...;
-
-// App sets up canvas on the page for use by the rendering sink.
-const canvas = document.getElementById("canvas");
+// The document contains a canvas for displaying VideoFrames.
+const canvasElement = document.getElementById("canvas");
+const canvasContext = canvasElement.getContext('bitmaprenderer');
 
-// Wrap the rendering sink in a writable stream so it can be connected to the graph.
-const renderingStream = new WritableStream(new CanvasRendererSink(canvas));
+function paintFrameToCanvas(videoFrame) {
+  // Paint every video frame ASAP for lowest latency.
+  canvasContext.transferFromImageBitmap(videoFrame.image);
+}
 
-// Construct the decoder and connect the pipes!
-const videoDecoder = new VideoDecoder({codec: 'vp8'});
-encodedVideo.pipeThrough(videoDecoder).pipeTo(renderingStream);
+const videoDecoder = new VideoDecoder({
+  output: paintFrameToCanvas,
+  error: onDecoderError
+});
 
-```
+videoDecoder.configure({codec: 'vp8'}).then((supported) => {
+  // More elaborate fallback recommended.
+  if (!supported)
```

Comment:
Unsupported is presumably Promise rejection.

Reply:
Sure, I've changed examples to work that way now. 

@mounirlamouri any opinion on whats more idiomatic? EME spec rejects when things aren't supported, while MC resolves with supported=false.

Reply:
Both EME and MC are correct. What matters is whether failure is expected. In the case of MC, we would expected unsupported configurations all the time but mis-configurations are being rejected. I would assume `configure()` to behave similarly: if the parameters passed in are incorrect, it should reject but if the actual device doesn't support the given configuration, it should probably just say not supported instead of rejecting.

Reply:
If the parameters cannot be supported then the codec is now in a bad state and must be discarded or recovered. Seems like it should be rejected to me.

Reply:
sounds good 👍